src/providers/azure.rs

```diff
@@ -0,0 +1,519 @@
+//! Both these types use [`Setup`] as their descriptor type.
+//!
+//! Azure does not support Spot or Defined Duration instances.
+//! As a result, if your tsunami crashes (i.e., exits without calling `drop()` on [`Launcher`], you must manually terminate your instances
+//! to avoid extra costs.
+//! The easiest way to do this is to delete resource groups beginning with `tsunami_`:
+//! `az group delete --name <name> --yes`.
+//! You can find such resource groups using:
+//! `az group list`.
+//!
+//! To use this provider, you must [install the Azure
+//! CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest),
+//! and the `az` command must be in your `$PATH`.
+//! Also, you must run `az login` to authenticate with Microsoft.
+//!
+//! # Example
+//! ```rust,no_run
+//! use tsunami::providers::{azure, Launcher};
+//! use tsunami::TsunamiBuilder;
+//!
+//! let mut b = TsunamiBuilder::default();
+//! b.add("my machine", azure::Setup::default()).unwrap();
+//! let mut l = azure::Launcher::default();
+//! b.spawn(&mut l).unwrap();
+//! let vms = l.connect_all().unwrap();
+//! let my_machine = vms.get("my machine").unwrap();
+//! let (stdout, stderr) = my_machine.ssh.as_ref().unwrap().cmd("echo \"Hello, Azure\"").unwrap();
+//! println!("{}", stdout);
+//! ```
+
+use crate::ssh;
+use educe::Educe;
+use failure::{bail, Error};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// A descriptor for a single Azure VM type.
+///
+/// The default is an UbuntuLTS, Standard_DS1_V2 VM in the East US region.
+#[derive(Clone, Educe)]
+#[educe(Debug)]
+pub struct Setup {
+    region: Region,
+    instance_type: String,
+    image: String,
+    username: String,
+    #[educe(Debug(ignore))]
+    setup_fn:
+        Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl Default for Setup {
+    fn default() -> Self {
+        Setup {
+            region: "eastus".parse().unwrap(),
+            instance_type: "Standard_DS1_v2".to_string(),
+            image: "UbuntuLTS".to_string(),
+            username: "ubuntu".to_string(),
+            setup_fn: None,
+        }
+    }
+}
+
+impl super::MachineSetup for Setup {
+    type Region = Region;
+
+    fn region(&self) -> Self::Region {
+        self.region
+    }
+}
+
+impl Setup {
+    /// See also [`Region`].
+    pub fn region(mut self, r: Region) -> Self {
+        self.region = r;
+        self
+    }
+
+    /// To view the available sizes in the relevant region, use:
+    /// `az vm list-sizes -l <region_name>`.
+    pub fn instance_type(mut self, inst_type: String) -> Self {
+        self.instance_type = inst_type;
+        self
+    }
+
+    /// Set the image.
+    ///
+    /// See `az vm image list` for valid options.
+    pub fn image(mut self, image: String) -> Self {
+        self.image = image;
+        self
+    }
+
+    pub fn username(mut self, username: String) -> Self {
```

Comment:
Docs (shouldn't this produce a warning?)

Reply:
Turns out `missing_docs` was turned off...